nui/src/main/java/org/terasology/nui/layouts/FlowLayout.java

```diff
@@ -50,41 +89,65 @@ public void removeAllWidgets() {
 
     @Override
     public void onDraw(Canvas canvas) {
-        int filledWidth = 0;
-        int filledHeight = 0;
-        int heightOffset = 0;
-        for (UIWidget widget : contents) {
-            Vector2i size = canvas.calculatePreferredSize(widget);
-            if (filledWidth != 0 && filledWidth + size.x  > canvas.size().x) {
-                heightOffset += filledHeight;
-                filledWidth = 0;
-                filledHeight = 0;
-            }
-            canvas.drawWidget(widget, RectUtility.createFromMinAndSize(filledWidth, heightOffset, size.x, size.y));
-            filledWidth += size.x;
-            filledHeight = Math.max(filledHeight, size.y);
-        }
+        layout(canvas, canvas.size(), true);
     }
 
     @Override
     public Vector2i getPreferredContentSize(Canvas canvas, Vector2i sizeHint) {
+        return layout(canvas, sizeHint, false);
+    }
+
+    /**
+     * Applies the flow layout to its children, arranging them in a directional flow.
+     * <p>
+     * The layout algorithm will wrap elements to a new line if adding them in the same line would exceed the bounding
+     * size. The first element of a row is placed regardless of its size, and thus may exceed the bounding size. The
+     * Flow layout attempts to stay within the preferred width, but will add new lines until all widgets have been laid
+     * out.
+     * <p>
+     * If the preferred bounding size is wider than the preferred size of the widest child the Flow layout guarantees to
+     * fit into the preferred width. There is no guarantee for the actual height.
+     *
+     * @param canvas the canvas to draw to and/or calculate sizes for
+     * @param boundingSize the boundary for this layout, may be a canvas size or a size hint
+     * @param draw whether to actually draw the widgets to the canvas
+     * @return the computed bounding box size for when arranging the widget
+     */
+    private Vector2i layout(Canvas canvas, Vector2i boundingSize, boolean draw) {
+        // current bounding box for the widgets already laid out
         Vector2i result = new Vector2i();
-        int filledWidth = 0;
-        int filledHeight = 0;
+        // current "cursor" position, always where the next widget is to be placed?
```

Comment:
> current "cursor" position, always where the next widget is to be placed?

is this still unclear or why is there a `?`?

Reply:
```suggestion
        // current horizontal "cursor" position, always where the next widget is to be placed?
```
or
```suggestion
        // current x-axis "cursor" position, always where the next widget is to be placed?
```

Reply:
This actually no longer holds with rtl support 🤔 should probably re-word this to make clear we're just talking about a size and not a position.